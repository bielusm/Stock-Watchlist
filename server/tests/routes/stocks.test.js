const { seedUser } = require('../fixtures/users');
const {
  seedDatabase,
  disconnect,
  connect
} = require('../fixtures/seedDatabase');

let token = '';

const request = require('supertest');
const moxios = require('moxios');
const app = require('../../app');

beforeAll(async () => {
  await connect();
  await seedDatabase();
});

afterAll(async () => {
  await disconnect();
});

beforeEach(async () => {
  await seedDatabase();
  res = await request(app)
    .post('/api/users/login')
    .send(seedUser);
  token = res.body.token;
  moxios.install();
});

afterEach(() => {
  moxios.uninstall();
});

describe('stocks route', () => {
  describe('test private routes', () => {
    test('should give proper error', async done => {
      const testRoutes = [
        { url: '/api/stocks/ibm', method: 'get', data: undefined }
      ];
      for (const route of testRoutes) {
        const { url, method, data } = route;
        let res = [];
        switch (method) {
          case 'get':
            res[0] = await request(app)
              .get(url)
              .send(data)
              .expect(400);
            res[1] = await request(app)
              .get(url)
              .set('x-auth-token', 'adhwdawdad')
              .send(data)
              .expect(401);
            break;
        }

        let errors = res[0].body.errors;
        expect(errors.length).toEqual(1);
        expect(errors[0].msg).toEqual('No Token In Header');

        errors = res[1].body.errors;
        expect(errors.length).toEqual(1);
        expect(errors[0].msg).toEqual('Not authorized');
      }
      done();
    });
  });

  describe('API frequency test', () => {
    test('returns propper warning', async done => {
      const testRoutes = [
        {
          url: '/api/stocks/ibm',
          method: 'get',
          data: undefined,
          apiUrl: `https://www.alphavantage.co/query?function=TIME_SERIES_WEEKLY&symbol=$ibm&apikey=${process.env.ALPHA_VANTAGE_KEY}`
        }
      ];
      for (route of testRoutes) {
        moxios.wait(() => {
          const request = moxios.requests.mostRecent();
          request.respondWith({
            status: 200,
            response: {
              Note:
                'Thank you for using Alpha Vantage! Our standard API call frequency is 5 calls per minute and 500 calls per day. Please visit https://www.alphavantage.co/premium/ if you would like to target a higher API call frequency.'
            }
          });
        });
        const { url, method, data } = route;
        let res;
        switch (method) {
          case 'get':
            res = await request(app)
              .get(url)
              .set('x-auth-token', token)
              .send(data)
              .expect(400);
            break;
        }
        let errors = res.body.errors;
        expect(errors.length).toEqual(1);
        expect(errors[0].msg).toEqual('Too many api calls');
        done();
      }
    });
  });

  describe('GET /api/stocks', () => {
    test('returns stock data', async done => {
      const { ALPHA_VANTAGE_KEY } = process.env;
      moxios.stubRequest(
        `https://www.alphavantage.co/query?function=TIME_SERIES_WEEKLY&symbol=ibm&apikey=${ALPHA_VANTAGE_KEY}`,
        {
          status: 200,
          response: require('../fixtures/stocks/ibmStocks.json')
        }
      );
      moxios.stubRequest(
        `https://www.alphavantage.co/query?function=GLOBAL_QUOTE&symbol=ibm&apikey=${ALPHA_VANTAGE_KEY}`,
        {
          status: 200,
          response: require('../fixtures/stocks/globalQuoteIbm.json')
        }
      );

      const response = await request(app)
        .get('/api/stocks/ibm')
        .set('x-auth-token', token)
        .expect(200);
      expect(response.body).toMatchSnapshot({});
      expect(response.body.ibm.last52).toHaveLength(52);
      done();
    });
    test('should return warning with bad symbol', async done => {
      const { ALPHA_VANTAGE_KEY } = process.env;
      moxios.stubRequest(
        `https://www.alphavantage.co/query?function=TIME_SERIES_WEEKLY&symbol=${'baddata'}&apikey=${ALPHA_VANTAGE_KEY}`,
        {
          status: 200,
          response: {
            'Error Message':
              'Invalid API call. Please retry or visit the documentation (https://www.alphavantage.co/documentation/) for TIME_SERIES_WEEKLY.'
          }
        }
      );
      const response = await request(app)
        .get('/api/stocks/baddata')
        .set('x-auth-token', token)
        .expect(400);
      expect(response.body).toEqual({
        errors: [{ msg: 'Invalid API call, possibly wrong symbol' }]
      });
      done();
    });
  });
});
