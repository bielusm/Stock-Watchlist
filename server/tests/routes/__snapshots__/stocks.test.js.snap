// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stocks route GET /api/stocks returns stock data 1`] = `
Object {
  "ibm": Object {
    "currentValue": 112.93,
    "last52": Array [
      Object {
        "date": "2020-03-27",
        "high": 113.15,
        "low": 90.56,
      },
      Object {
        "date": "2020-03-20",
        "high": 107.41,
        "low": 94.71,
      },
      Object {
        "date": "2020-03-13",
        "high": 124.88,
        "low": 100.81,
      },
      Object {
        "date": "2020-03-06",
        "high": 136.095,
        "low": 124.52,
      },
      Object {
        "date": "2020-02-28",
        "high": 148.05,
        "low": 126.36,
      },
      Object {
        "date": "2020-02-21",
        "high": 151.89,
        "low": 148.84,
      },
      Object {
        "date": "2020-02-14",
        "high": 155.6,
        "low": 149.77,
      },
      Object {
        "date": "2020-02-07",
        "high": 158.75,
        "low": 143.9,
      },
      Object {
        "date": "2020-01-31",
        "high": 144.05,
        "low": 134.97,
      },
      Object {
        "date": "2020-01-24",
        "high": 145.79,
        "low": 137.6,
      },
      Object {
        "date": "2020-01-17",
        "high": 138.33,
        "low": 135.07,
      },
      Object {
        "date": "2020-01-10",
        "high": 137.87,
        "low": 133.2,
      },
      Object {
        "date": "2020-01-03",
        "high": 135.92,
        "low": 132.4,
      },
      Object {
        "date": "2019-12-27",
        "high": 136.15,
        "low": 134.61,
      },
      Object {
        "date": "2019-12-20",
        "high": 136.42,
        "low": 133.46,
      },
      Object {
        "date": "2019-12-13",
        "high": 135.6604,
        "low": 133.26,
      },
      Object {
        "date": "2019-12-06",
        "high": 134.5,
        "low": 130.689,
      },
      Object {
        "date": "2019-11-29",
        "high": 136.15,
        "low": 133.03,
      },
      Object {
        "date": "2019-11-22",
        "high": 135.38,
        "low": 132.75,
      },
      Object {
        "date": "2019-11-15",
        "high": 137.2,
        "low": 133.33,
      },
      Object {
        "date": "2019-11-08",
        "high": 139.135,
        "low": 136.16,
      },
      Object {
        "date": "2019-11-01",
        "high": 136.63,
        "low": 133.2,
      },
      Object {
        "date": "2019-10-25",
        "high": 135.93,
        "low": 130.9,
      },
      Object {
        "date": "2019-10-18",
        "high": 143.72,
        "low": 132.25,
      },
      Object {
        "date": "2019-10-11",
        "high": 144.5,
        "low": 138.25,
      },
      Object {
        "date": "2019-10-04",
        "high": 147.35,
        "low": 139.18,
      },
      Object {
        "date": "2019-09-27",
        "high": 145.09,
        "low": 140.558,
      },
      Object {
        "date": "2019-09-20",
        "high": 145.03,
        "low": 140.51,
      },
      Object {
        "date": "2019-09-13",
        "high": 145.46,
        "low": 140.46,
      },
      Object {
        "date": "2019-09-06",
        "high": 141.695,
        "low": 133.3301,
      },
      Object {
        "date": "2019-08-30",
        "high": 136.07,
        "low": 129.16,
      },
      Object {
        "date": "2019-08-23",
        "high": 136.33,
        "low": 128.83,
      },
      Object {
        "date": "2019-08-16",
        "high": 136.57,
        "low": 130.25,
      },
      Object {
        "date": "2019-08-09",
        "high": 145.1,
        "low": 135.35,
      },
      Object {
        "date": "2019-08-02",
        "high": 152.95,
        "low": 145.59,
      },
      Object {
        "date": "2019-07-26",
        "high": 151.94,
        "low": 148.38,
      },
      Object {
        "date": "2019-07-19",
        "high": 151.58,
        "low": 141.95,
      },
      Object {
        "date": "2019-07-12",
        "high": 142.92,
        "low": 139.13,
      },
      Object {
        "date": "2019-07-05",
        "high": 141.82,
        "low": 139.28,
      },
      Object {
        "date": "2019-06-28",
        "high": 140.15,
        "low": 137.84,
      },
      Object {
        "date": "2019-06-21",
        "high": 139.54,
        "low": 134.7,
      },
      Object {
        "date": "2019-06-14",
        "high": 136.46,
        "low": 133.91,
      },
      Object {
        "date": "2019-06-07",
        "high": 134.72,
        "low": 127.06,
      },
      Object {
        "date": "2019-05-31",
        "high": 132.64,
        "low": 126.85,
      },
      Object {
        "date": "2019-05-24",
        "high": 137.06,
        "low": 130.44,
      },
      Object {
        "date": "2019-05-17",
        "high": 136.11,
        "low": 130.96,
      },
      Object {
        "date": "2019-05-10",
        "high": 140.69,
        "low": 132.42,
      },
      Object {
        "date": "2019-05-03",
        "high": 141.81,
        "low": 138.68,
      },
      Object {
        "date": "2019-04-26",
        "high": 141.31,
        "low": 137.71,
      },
      Object {
        "date": "2019-04-18",
        "high": 145.39,
        "low": 136.26,
      },
      Object {
        "date": "2019-04-12",
        "high": 144.44,
        "low": 141.84,
      },
      Object {
        "date": "2019-04-05",
        "high": 144.22,
        "low": 141.51,
      },
    ],
    "stats": Object {
      "highest": 158.75,
      "lowest": 90.56,
    },
  },
}
`;
